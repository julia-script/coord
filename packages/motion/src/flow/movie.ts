import {
  MotionBuilder,
  MotionScene,
  isEndRequest,
  isRequest,
  makeState,
  requestContext,
  requestEnd,
  runMotion,
} from "@/motion";
import { InferThread, Thread } from "@/utils";
import { YieldedType, raise } from "@coord/core";

export function* scene<
  TKey extends string,
  TScene extends MotionScene<MotionBuilder>
>(key: TKey, scene: TScene, transition = 0) {
  const parentContext = yield* requestContext();

  const context = runMotion(scene, {
    ...parentContext.settings,
    transitionDuration: transition,
  });

  context.meta.start =
    parentContext.frames.length;
  parentContext.meta.scenes[key] = context.meta;
  const firstFrame = context.frames[0] ?? raise();
  const sceneState = yield* makeState<
    TKey,
    typeof firstFrame
  >(key, firstFrame);
  let hasRequestedNext = false;
  for (
    let i = 0;
    i < context.frames.length;
    i++
  ) {
    const frame = context.frames[i] ?? raise();

    if (!hasRequestedNext && frame.$done) {
      hasRequestedNext = true;
      yield* requestEnd();
    }

    yield sceneState.set(frame);
  }
  if (!hasRequestedNext) {
    yield* requestEnd();
  }
  sceneState.set((curr) => ({
    ...curr,
    $done: true,
  }));
}

export function movie<TThread extends Thread[]>(
  ...threads: TThread
) {
  return function* () {
    const activeScenes = new Set<
      TThread[number]
    >();
    const startNext = () => {
      const next = threads.shift();
      if (!next) return;
      activeScenes.add(next);
    };
    startNext();
    while (true) {
      for (const iterable of activeScenes) {
        while (true) {
          const { done, value } = iterable.next();
          if (isRequest(value)) {
            if (isEndRequest(value)) {
              startNext();
              continue;
            }
            yield value as YieldedType<
              InferThread<TThread[number]>
            >;
            continue;
          }
          if (done) {
            activeScenes.delete(iterable);
          }
          break;
        }
      }
      if (!activeScenes.size) break;

      yield;
    }
  };
}
