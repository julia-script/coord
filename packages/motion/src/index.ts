export * from "./player";
export * from "./utils";
export * from "./flow";
export * from "./tweening";
export * from "./controls";
export * from "./effects";
export * from "./motion";
