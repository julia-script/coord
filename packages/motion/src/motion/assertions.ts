import {
  YieldedType,
  isObject,
} from "@coord/core";
import {
  ContextRequest,
  EndRequest,
  MakeStateRequest,
  MotionRequest,
  MotionState,
} from "./types";

export const isRequest = (
  value: unknown
): value is MotionRequest => {
  return isObject(value) && "type" in value;
};

export const isContextRequest = <
  TState extends MotionState
>(
  value: unknown
): value is YieldedType<
  ContextRequest<TState>
> => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_CONTEXT"
  );
};

export const isMakeStateRequest = <
  TKey extends string,
  TValue
>(
  value: unknown
): value is MakeStateRequest<TKey, TValue> => {
  return (
    isRequest(value) &&
    value.type === "MAKE_STATE" &&
    "state" in value &&
    isObject(value.state) &&
    "key" in value.state &&
    "initialState" in value.state
  );
};

export const isEndRequest = (
  value: unknown
): value is EndRequest => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_END"
  );
};
