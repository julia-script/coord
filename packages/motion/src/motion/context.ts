import {
  isContextRequest,
  isEndRequest,
  isMakeStateRequest,
  isRequest,
} from "./assertions";
import {
  MotionBuilder,
  MotionMeta,
  MotionRequest,
  MotionScene,
  MotionSettings,
  MotionState,
  SceneState,
} from "./types";

type FullState<TState extends MotionState> = TState & {
  $transition: number;
  $frame: number;
  $done: boolean;
};
type FullStateKeys<TState extends MotionState> = keyof FullState<TState> &
  string;

export class MotionContext<TState extends MotionState> {
  _state: FullState<TState>;
  frames: FullState<TState>[] = [];
  meta: MotionMeta = {
    title: "Untitled",
    start: 0,
    duration: 0,
    scenes: {},
  };

  settings: MotionSettings = {
    fps: 60,
    physicsFps: 120,
  };

  get time() {
    return this.frames.length / this.settings.fps;
  }
  constructor() {
    this._state = {} as FullState<TState>;
  }

  static from<TScene extends MotionScene<MotionBuilder>>(
    scene: TScene,
    settings: Partial<MotionSettings> & { transitionDuration?: number } = {}
  ): MotionContext<SceneState<TScene>> {
    const context = new MotionContext<SceneState<TScene>>();
    context.meta.title = scene.meta.title;
    context.meta.description = scene.meta.description;

    const { transitionDuration = 0, ...rest } = settings;
    context.settings = { ...context.settings, ...rest };
    const builderIterator = scene.builder();

    let transitionDurationInFrames = context.settings.fps * transitionDuration;
    context._state.$done = false;

    let done = false;
    while (true) {
      if (done === false) {
        const currentIteration = builderIterator.next();
        done = !!currentIteration.done;
        if (isRequest(currentIteration.value)) {
          context.respondRequest(currentIteration.value);
          continue;
        }
      }
      if (
        done === true &&
        context.frames.length >= transitionDurationInFrames
      ) {
        break;
      }

      context._state.$frame = context.frames.length;

      context._state.$transition =
        transitionDurationInFrames <= 0
          ? 1
          : Math.min(
              (context.frames.length + 1) / transitionDurationInFrames,
              1
            );

      context.pushFrame();
    }
    if (context.frames.length === 0) {
      context._state.$done = true;
      context.pushFrame();
    }
    return context;
  }

  pushFrame() {
    this.meta.duration++;
    this.frames.push({
      ...this._state,
    });
  }

  setState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: (typeof this._state)[TKey]
  ) {
    this._state[key] = value;
  }

  respondRequest<TRequest extends MotionRequest>(request: TRequest) {
    if (isContextRequest<TState>(request)) {
      request.context = this;
    } else if (
      isMakeStateRequest<
        FullStateKeys<TState>,
        FullState<TState>[FullStateKeys<TState>]
      >(request)
    ) {
      this.makeState(request.state.key, request.state.initialState);
    } else if (isEndRequest(request)) {
      this._state.$done = true;
    }
  }

  makeState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: FullState<TState>[TKey]
  ) {
    this.setState(key, value);

    for (let i = this.frames.length - 1; i >= 0; i--) {
      const frame = this.frames[i]!;
      if (key in frame) {
        break;
      }
      Object.assign(frame, { [key]: value });
    }
  }
}
