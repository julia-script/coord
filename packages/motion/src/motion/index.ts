export * from "./assertions";
export * from "./context";
export * from "./requests";
export * from "./types";
export * from "./make-scene";
export * from "./run-motion";
