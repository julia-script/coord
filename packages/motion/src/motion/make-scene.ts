import { SceneMetaish, generateMeta } from "@/utils";
import { MotionBuilder } from "./types";

export function makeMotion<TBuilder extends MotionBuilder>(
  meta: SceneMetaish,
  builder: TBuilder
) {
  return {
    meta: generateMeta(meta),
    builder,
  };
}
