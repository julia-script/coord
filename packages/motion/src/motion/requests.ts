import { isFunction } from "@coord/core";
import { MotionState } from "./types";
import type { MotionContext } from "./context";
import { control } from "@/controls/control";

export function* requestContext<
  TState extends MotionState
>() {
  const request: {
    type: "REQUEST_CONTEXT";
    context?: MotionContext<TState>;
  } = {
    type: "REQUEST_CONTEXT",
  };

  yield request;

  if (!request.context) {
    throw new Error("Context is not provided");
  }

  return request.context;
}

export function* requestEnd() {
  yield { type: "REQUEST_END" } as const;
}

export function* makeState<
  TKey extends string,
  TValue
>(key: TKey, initialState: TValue) {
  const context = yield* requestContext();
  yield {
    type: "MAKE_STATE",
    state: { key, initialState },
  } as const;

  const getter = () =>
    context._state[key] as TValue;
  const setter = (
    value: TValue | ((prev: TValue) => TValue)
  ) => {
    context._state[key] = isFunction(value)
      ? value(getter())
      : value;
  };
  return control(getter, setter);
}
