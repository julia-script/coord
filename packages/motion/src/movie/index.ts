export * from "./make-movie";
export * from "./make-scene";
