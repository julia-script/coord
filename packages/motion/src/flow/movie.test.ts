import { makeState, requestEnd, runMotion } from "@/motion";
import { makeMotion } from "@/motion";
import { test, describe, expect } from "vitest";

import { movie, scene } from "./movie";

describe("movie", async () => {
  test("without transition", async () => {
    const scene1 = makeMotion("Test A", function* () {
      const test = yield* makeState("a", 0);
      yield test.set(1);
      yield test.set(2);
    });
    const scene2 = makeMotion("Test B", function* () {
      const test = yield* makeState("b", 0);
      yield test.set(1);
      yield test.set(2);
    });
    const scene3 = makeMotion("Test C", function* () {
      const test = yield* makeState("c", 0);
      yield test.set(1);
      yield test.set(2);
    });

    const testMovie = makeMotion(
      "Test",
      movie(
        scene("scene1", scene1),
        scene("scene2", scene2),
        scene("scene3", scene3)
      )
    );
    const executed = runMotion(testMovie, {
      fps: 1,
    });
    expect(executed.frames).toMatchSnapshot();
  });

  test("with transition", async () => {
    const scene1 = makeMotion("Test A", function* () {
      const test = yield* makeState("a", 0);
      yield* requestEnd();
      yield test.set(1);
      yield test.set(2);
    });
    const scene2 = makeMotion("Test B", function* () {
      const test = yield* makeState("b", 0);
      yield test.set(1);
      yield test.set(2);
    });

    const testMovie = makeMotion(
      "Test",
      movie(scene("scene1", scene1), scene("scene2", scene2))
    );
    const executed = runMotion(testMovie, {
      fps: 1,
    });

    expect(executed.frames).toMatchSnapshot();
  });
});
