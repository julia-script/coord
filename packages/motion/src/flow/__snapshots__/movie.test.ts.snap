// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`movie > with transition 1`] = `
[
  {
    "$done": false,
    "$frame": 0,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 0,
      "$transition": 1,
      "a": 1,
    },
    "scene2": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "b": 1,
    },
  },
  {
    "$done": false,
    "$frame": 1,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": false,
      "$frame": 1,
      "$transition": 1,
      "b": 2,
    },
  },
]
`;

exports[`movie > without transition 1`] = `
[
  {
    "$done": false,
    "$frame": 0,
    "$transition": 1,
    "scene1": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "a": 1,
    },
    "scene2": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "b": 1,
    },
    "scene3": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "c": 1,
    },
  },
  {
    "$done": false,
    "$frame": 1,
    "$transition": 1,
    "scene1": {
      "$done": false,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "b": 1,
    },
    "scene3": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "c": 1,
    },
  },
  {
    "$done": false,
    "$frame": 2,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "b": 1,
    },
    "scene3": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "c": 1,
    },
  },
  {
    "$done": false,
    "$frame": 3,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": false,
      "$frame": 1,
      "$transition": 1,
      "b": 2,
    },
    "scene3": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "c": 1,
    },
  },
  {
    "$done": false,
    "$frame": 4,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "b": 2,
    },
    "scene3": {
      "$done": false,
      "$frame": 0,
      "$transition": 1,
      "c": 1,
    },
  },
  {
    "$done": false,
    "$frame": 5,
    "$transition": 1,
    "scene1": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "a": 2,
    },
    "scene2": {
      "$done": true,
      "$frame": 1,
      "$transition": 1,
      "b": 2,
    },
    "scene3": {
      "$done": false,
      "$frame": 1,
      "$transition": 1,
      "c": 2,
    },
  },
]
`;
